File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int N = Integer.parseInt(br.readLine());
        for(int i=0; i<N; i++) {
            char[] str = br.readLine().toCharArray();
            Stack<Integer> stack = new Stack<>();
            boolean isVPS = true;
            for(char c : str) {
                if(c == '(') stack.push(1);
                else if(!stack.isEmpty()) stack.pop();
                else {
                    isVPS = false;
                    break;
                }
            }
            sb.append(stack.isEmpty() && isVPS ? "YES" : "NO").append("\n");
        }
        System.out.println(sb);
    }
}

File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>
#define fastio ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

using namespace std;

int n;

int main(void) {
    fastio;
    cin >> n;
    for (int i = 0; i < n; i++) {
        string str;
        cin >> str;
        stack<int> st;
        bool isVPS = true;
        for (auto c : str) {
            if (c == '(') st.push(1);
            else if (!st.empty()) st.pop();
            else {
                isVPS = false;
                break;
            }
        }
        if (st.empty() && isVPS) cout << "YES\n";
        else cout << "NO\n";
    }

    return 0;
}